### ISSUE TITLE
JSDoc: inference and type checking breaks when param is destructured

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#16219]
**TypeScript Version:**  2.3.4

**Code**

This works:

```ts
{
  /**
   * @typedef {{
      type: string,
      color?: string
    }} Props
  * @param {Props} props
  */
  const foo = (props) => {
    console.log(props.type); // type of type is string
  };
  foo({ type: 1 }); // error

  // However, if I destructure the param, it breaks:

  /**
   * @typedef {{
      type: string,
      color?: string
    }} Props2
  * @param {Props2} props
  */
  const foo2 = ({ type }) => {
    console.log(type); // type of type is any, expected string
  };
  foo2({ type: 1 }); // expected error but got none
}
```

Reading through [the docs on JSDoc support](https://github.com/Microsoft/TypeScript/wiki/JSDoc-support-in-JavaScript), I couldn't find anything to suggest why this shouldn't work.

**Comment by OliverJAsh** [comment]
Possibly related to https://github.com/Microsoft/TypeScript/issues/11859

**Comment by mhegazy** [comment]
should be covered by https://github.com/Microsoft/TypeScript/issues/11859, though i think the syntax in jsdoc comments in this report are the correct ones.

**Comment by mhegazy** [comment]
on a second thought, should keep them separate, the two scenarios are different enough that we may want to fix one and not the other.

**Comment by sandersn** [comment]
Fixed, I believe by #23307
